exonum-java-binding/core/src/main/java/com/exonum/binding/core/storage/database/Fork.java

```diff
@@ -58,21 +66,90 @@ public static Fork newInstance(long nativeHandle, boolean owningHandle, Cleaner
 
     NativeHandle h = new NativeHandle(nativeHandle);
     // Add an action destroying the native peer if necessary.
-    ProxyDestructor.newRegistered(cleaner, h, Fork.class, nh -> {
+    ProxyDestructor destructor = ProxyDestructor.newRegistered(cleaner, h, Fork.class, nh -> {
       if (owningHandle) {
         Views.nativeFree(nh);
       }
     });
 
-    return new Fork(h, cleaner);
+    // Create a cleaner for collections. A separate cleaner is needed to be able to destroy
+    // the objects depending on the fork when it is converted into patch and invalidated
+    Cleaner forkCleaner = new Cleaner();
+    cleaner.add(forkCleaner::close);
+
+    return new Fork(h, destructor, forkCleaner);
   }
 
   /**
    * Create a new owning Fork.
    *
    * @param nativeHandle a handle of the native Fork object
+   * @param destructor a destructor of the native peer, registered with the parent cleaner
+   * @param forkCleaner a cleaner for objects depending on the fork
+   */
+  private Fork(NativeHandle nativeHandle, ProxyDestructor destructor, Cleaner forkCleaner) {
+    super(nativeHandle, forkCleaner, new IncrementalModificationCounter(), true);
+    this.destructor = destructor;
+  }
+
+  /**
+   * Converts this fork into a patch that can be merged into the database.
```

Comment:
Provide a link to patch in documentation when it's ready? https://exonum.com/doc/version/latest/architecture/merkledb/#patches

Reply:
It is an internal method that depends on the native method producing the patch, therefore, this documentation is for us. It is probably faster to navigate to the Patch declaration in the native code which might have better documentation that the website, but the link won't hurt.